### PR TITLE
feat: DeepseekV4 onboard to SGLang DLC

### PR DESCRIPTION
**Pull request by sirutBuasai**
# SGLang Benchmark Report — DeepSeek V4 Pro Multi-Node

## Setup

| Parameter | Value |
|-----------|-------|
| Model | DeepSeek V4 Pro (1.6T MoE, 49B active params, FP4) |
| Nodes | 2x p5.48xlarge (8x H100 80GB each) |
| TP | 16 |
| Interconnect | EFA (32 interfaces per node, RDMA, no placement group) |
| Container | `<dev_account.dkr.ecr.<dev_region>.amazonaws.com/<dev_repo>:sglang-0.5.12-gpu-py312-cu129-amzn2023-ec2` |
| SGLang flags | `--moe-runner-backend marlin --mem-fraction-static 0.9 --context-length 4096 --trust-remote-code` |
| NCCL env | `NCCL_SOCKET_IFNAME=enp71s0 FI_EFA_USE_DEVICE_RDMA=1 NCCL_IB_DISABLE=1 FI_PROVIDER=efa SGLANG_SHARED_EXPERT_TP1=1` |
| Account | (dev_account) |
| ODCR | (dev_odcr) |
| Date | 2026-06-01 |

## Throughput

| Model | Runner | TP | Input Len | Output Len | Prompts | Output Tok/s | Total Tok/s | Requests/s | Elapsed (s) |
|-------|--------|----|-----------|------------|---------|--------------|-------------|------------|-------------|
| deepseek-v4-flash | gpu-h100-8gpu-runners | 8 | 512 | 128 | 32 | 1284.03 | 6890.53 | 20.12 | 1.59 |
| deepseek-v4-pro | 2x p5.48xlarge (EFA) | 16 | 512 | 128 | 32 | 42.93 | 230.38 | 0.67 | 47.56 |

## Latency (Online Serving)

| Model | Runner | TP | Mean TTFT (ms) | p99 TTFT (ms) | Mean TPOT (ms) | p99 TPOT (ms) | Mean ITL (ms) | p99 ITL (ms) |
|-------|--------|----|----------------|---------------|----------------|---------------|---------------|--------------|
| deepseek-v4-flash | gpu-h100-8gpu-runners | 8 | 362.34 | 366.30 | 10.56 | 11.23 | 10.60 | 11.95 |
| deepseek-v4-pro | 2x p5.48xlarge (EFA) | 16 | 23634.73 | 45953.97 | 22.44 | 22.75 | 22.45 | 23.13 |

## Analysis

- **TPOT is 2x slower** (22ms vs 10.5ms) — expected since TP=16 across 2 nodes has more all-reduce communication overhead than TP=8 on one node via NVLink
- **TTFT is very high** (23.6s) — the server processes 1 request at a time due to SWA (sliding window attention) cache saturation at just 1 running request. Prefill for 512 tokens takes ~750ms, but with 32 requests queued sequentially it takes 32 × ~750ms ≈ 24s
- **Output throughput: 43 tok/s** — limited by the inability to batch multiple requests concurrently. Peak is 92 tok/s (single request decode speed)
- **KV cache is the bottleneck** — with `--mem-fraction-static 0.9`, the budget is 72GB per GPU. Model weights take ~46.4GB, leaving only ~25.6GB for KV cache. The SWA cache saturates at 1 concurrent request.
- **EFA is working correctly** — NCCL confirmed `Selected Provider is efa (found 32 nics)` with RDMA transport. The multi-node NCCL all_reduce_perf test showed 164 GB/s at 128MB (full expected bandwidth).

## Reference: SGLang Cookbook Numbers

| Source | GPU | TP | Config | Output Tok/s |
|--------|-----|----|----|--------------|
| SGLang cookbook (max throughput) | H100 | 16 (2 nodes) | `--mem-fraction-static 0.9` | Not published |
| SGLang team (reported) | B200 | 8 (1 node) | 4096 in / 4096 out | 198 tok/s |
| Our result | H100 | 16 (2 nodes) | 512 in / 128 out | 42.93 tok/s |

## Key Findings from This Test

1. **EFA multi-node works on raw EC2** — no placement group required, just correct security group (self-referencing egress rule) + `NCCL_SOCKET_IFNAME`
2. **Security group egress was the previous blocker** — adding all-traffic self-referencing egress fixed the "Unreachable remote" NCCL error
3. **ODCR + same AZ is sufficient** — instances from different ODCRs in the same AZ can communicate over EFA
4. **DeepGEMM JIT warmup is critical** — first run after server start shows ~10 tok/s due to kernel compilation; subsequent runs show 43 tok/s
5. **KV cache limits batching** — the V4 Pro model uses ~46.4GB per GPU for weights (FP4), leaving minimal room for KV cache on H100 80GB

## Test Result

[Deepseek v4 Flash](https://github.com/aws/deep-learning-containers/actions/runs/26731434223)
Deepseek v4 Pro ran with local cluster without CI because multi-node testing in CI is not set up yet.

______________________________________________________________________

<details>
<summary>Toggle if you are merging into master Branch</summary>

By default, docker image builds and tests are disabled. Two ways to run builds and tests:

1. Using dlc_developer_config.toml
1. Using this PR description (currently only supported for PyTorch, TensorFlow, vllm, and base images)

<details>
<summary>How to use the helper utility for updating dlc_developer_config.toml</summary>

Assuming your remote is called `origin` (you can find out more with `git remote -v`)...

- Run default builds and tests for a particular buildspec - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -cp origin`

- Enable specific tests for a buildspec or set of buildspecs - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -t sanity_tests -cp origin`

- Restore TOML file when ready to merge

`python src/prepare_dlc_dev_environment.py -rcp origin`

**NOTE: If you are creating a PR for a new framework version, please ensure success of the local, standard, rc, and efa sagemaker tests by updating the dlc_developer_config.toml file:**

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`
- [ ] `sagemaker_local_tests = true`

</details>

<details>
<summary>How to use PR description</summary>
Use the code block below to uncomment commands and run the PR CodeBuild jobs. There are two commands available:

- `# /buildspec <buildspec_path>`
  - e.g.: `# /buildspec pytorch/training/buildspec.yml`
  - If this line is commented out, dlc_developer_config.toml will be used.
- `# /tests <test_list>`
  - e.g.: `# /tests sanity security ec2`
  - If this line is commented out, it will run the default set of tests (same as the defaults in dlc_developer_config.toml): `sanity, security, ec2, ecs, eks, sagemaker, sagemaker-local`.

</details>

```
# /buildspec <buildspec_path>
# /tests <test_list>
```

</details>

<details>
<summary>Toggle if you are merging into main Branch</summary>

## PR Checklist

- [] I ran `pre-commit run --all-files` locally before creating this PR. (Read [DEVELOPMENT.md](https://github.com/aws/deep-learning-containers/blob/main/DEVELOPMENT.md) for details).

</details>

### CODE REVIEW

File: .github/config/image/sglang-sagemaker-amzn2023.yml [hunk 1]
```diff
@@ -1,10 +1,12 @@
 # SGLang SageMaker AL2023 Image Configuration
+# Single source of truth for image identity, build pins, and release settings.
+# Workflow extracts both image tags and --build-args from this file.
 image:
   name: "sglang-sagemaker-amzn2023"
   description: "SGLang for SageMaker on Amazon Linux 2023"
 common:
   framework: "sglang"
-  framework_version: "0.5.10"
+  framework_version: "0.5.12"
```
Comment (sallyseok) [inline .github/config/image/sglang-sagemaker-amzn2023.yml:9]: Why is this one not `"0.5.12.dev543"`?

File: .github/workflows/prcheck-merge-conditions.yml [hunk 2]
```diff
@@ -26,7 +26,6 @@ jobs:
             const self = 'enforce-all-checks';
             const ignoredChecks = new Set([self]);
             const ignoredPatterns = [
-              /^endpoint-test \/ endpoint-test/,  // SageMaker endpoint tests — non-blocking
```
Comment (sallyseok) [inline .github/workflows/prcheck-merge-conditions.yml:29]: This is because we don't run into capacity issues anymore right

Comment (sirutBuasai) [reply .github/workflows/prcheck-merge-conditions.yml:29]: yup, I actually changed sglang endpoint instance from g5 to g6 which is less likely to encounter capacity issue as well.

File: .github/config/image/sglang-sagemaker-amzn2023.yml (hunk 1 above)
Comment (sirutBuasai) [reply .github/config/image/sglang-sagemaker-amzn2023.yml:9]: I overlooked it lol. But honestly this version doesn't matter anymore for amzn2023 images since this framework_version only determine the image tag in CI. In prod, it will be `sglang:server-cuda-v1`.

I will fix this field in another PR where I will create the autorelease workflow for sglang though. I don't wanna rerun the whole CI tests from this one change.

File: test/sglang/scripts/sglang_model_smoke_test.sh [hunk 3]
```diff
@@ -9,8 +9,8 @@ MODEL_NAME="${2:?Usage: $0 <model_dir> <model_name> [extra_args...]}"
 shift 2
 EXTRA_ARGS="$*"
 
-SGLANG_PORT=8000
-HEALTH_TIMEOUT=600
+SGLANG_PORT="${SGLANG_PORT:-30000}"
```
Comment (sallyseok) [inline test/sglang/scripts/sglang_model_smoke_test.sh:12]: why the port change?

**Review by sallyseok — APPROVED** [review]

File: test/sglang/scripts/sglang_model_smoke_test.sh (hunk 3 above)
Comment (sirutBuasai) [reply test/sglang/scripts/sglang_model_smoke_test.sh:12]: SGLang default port is 30000, https://sgl-project.github.io/basic_usage/send_request.html, we should keep inline with them. vllm default port is 8000.

The only edge case is that sagemaker endpoint uses port 8080 for healthchecks so the sagemaker_entrypoint still defaults to 8080.
